src/brod_client.erl

```diff
@@ -981,6 +1000,74 @@ ensure_partition_workers(TopicName, State, F) ->
       end
     end).
 
+-spec maybe_start_partition_producer(TopicsMetadata, Producers, Topics, ProducerSup, Configs) -> ok
+when
+  TopicsMetadata :: #{Key => non_neg_integer()},
+  Producers :: #{Key => non_neg_integer()},
+  Topics :: [topic()],
+  ProducerSup :: pid(),
+  Configs :: #{Key => brod_producer:config()},
+  Key :: topic().
+maybe_start_partition_producer(_TopicsMetadata, _Producers, [], _ProducerSuf, _Configs) ->
+  ok;
+maybe_start_partition_producer(TopicsMetadata, Producers, [Topic | Rest], ProducerSup, Configs) when
+  is_map_key(Topic, TopicsMetadata), is_map_key(Topic, Producers)
+->
+  OldPartitionCnt = maps:get(Topic, Producers),
+  NewPartitionCnt = maps:get(Topic, TopicsMetadata),
+  lists:foreach(
+    fun
+      (PartitionIdx) when is_integer(PartitionIdx) ->
+        sync_partition(ProducerSup, Topic, PartitionIdx, Configs);
+      (_) ->
+        ok
+    end,
+    lists:seq(OldPartitionCnt, OldPartitionCnt + (NewPartitionCnt - OldPartitionCnt) - 1)
+  ),
+
+  maybe_start_partition_producer(TopicsMetadata, Producers, Rest, ProducerSup, Configs);
+maybe_start_partition_producer(TopicsMetadata, Producers, [_Topic | Rest], ProducerSup, Configs) ->
+  maybe_start_partition_producer(TopicsMetadata, Producers, Rest, ProducerSup, Configs).
+
+-spec filter_topics([kpro:struct()]) -> #{topic() => non_neg_integer()}.
+filter_topics(TopicsMetadataArray) ->
+  {NoErrors1, _} =
+    lists:partition(fun(#{error_code := E}) -> E =:= no_error end, TopicsMetadataArray),
+
+  NoErrors = [
+    {TopicName, erlang:length(Partitions)}
+    || #{name := <<TopicName/binary>>, partitions := Partitions} <- NoErrors1,
+    is_list(Partitions)
+  ],
+
+  maps:from_list(NoErrors).
+
+-spec started_producers(pid()) -> #{topic() => non_neg_integer()}.
+started_producers(ProducerSup) ->
+  Producers = brod_supervisor3:which_children(ProducerSup),
```

Comment:
`brod_supervisor3:which_children(ProducerSup)` will give you the per-topic supervisors which supervises the per-partition workers

i think you'll need to do it like this:

```
started_producers(ProducerSup) ->
  TopicSups = which_producers(ProducerSup),
  lists:foldl(
    fun({Topic, Pid, _Type, _Mods}, Acc) when is_pid(Pid) ->
      PartitionWorkers = which_producers(Pid),
      case length(PartitionWorkers) of
        0 -> Acc;
        L -> Acc#{Topic => L}
    (_) ->
      Acc
    end, TopicSups).
```
and this
```
which_producers(Sup) ->
  try
    brod_supervisor3:which_children(Sup)
  catch
    _:_ ->
     []
  end.
```